crates/oxc_transformer/src/lib.rs

```diff
@@ -89,41 +75,50 @@ impl<'a> Transformer<'a> {
         trivias: Trivias,
         options: TransformOptions,
     ) -> Self {
-        let ctx = Rc::new(TransformCtx::new(
-            allocator,
-            source_path,
-            source_type,
-            source_text,
-            trivias,
-            &options,
-        ));
-        Self {
-            ctx: Rc::clone(&ctx),
-            x0_typescript: TypeScript::new(options.typescript, Rc::clone(&ctx)),
-            x1_react: React::new(options.react, Rc::clone(&ctx)),
-            x2_es2021: ES2021::new(options.es2021),
-            x2_es2020: ES2020::new(options.es2020),
-            x2_es2019: ES2019::new(options.es2019),
-            x2_es2018: ES2018::new(options.es2018),
-            x2_es2016: ES2016::new(options.es2016),
-            x3_es2015: ES2015::new(options.es2015),
-            x4_regexp: RegExp::new(options.regexp, ctx),
-        }
+        let ctx =
+            TransformCtx::new(allocator, source_path, source_type, source_text, trivias, &options);
+        Self { ctx, options }
     }
 
     pub fn build_with_symbols_and_scopes(
-        mut self,
+        self,
         symbols: SymbolTable,
         scopes: ScopeTree,
         program: &mut Program<'a>,
     ) -> TransformerReturn {
         let allocator = self.ctx.ast.allocator;
-        let (symbols, scopes) = traverse_mut(&mut self, allocator, program, symbols, scopes);
+
+        let mut transformer = TransformerImpl {
+            x0_typescript: TypeScript::new(self.options.typescript, &self.ctx),
+            x1_react: React::new(self.options.react, &self.ctx),
+            x2_es2021: ES2021::new(self.options.es2021),
+            x2_es2020: ES2020::new(self.options.es2020),
+            x2_es2019: ES2019::new(self.options.es2019),
+            x2_es2018: ES2018::new(self.options.es2018),
+            x2_es2016: ES2016::new(self.options.es2016),
+            x3_es2015: ES2015::new(self.options.es2015),
+            x4_regexp: RegExp::new(self.options.regexp, &self.ctx),
+        };
+
+        let (symbols, scopes) = traverse_mut(&mut transformer, allocator, program, symbols, scopes);
         TransformerReturn { errors: self.ctx.take_errors(), symbols, scopes }
     }
 }
 
-impl<'a> Traverse<'a> for Transformer<'a> {
+struct TransformerImpl<'a, 'c> {
```

Comment:
🤮 Should we use `'ast` and `'ctx` just like some other hardcore projects?

Reply:
I've pushed a commit to change `'c` to `'ctx`.

I'm less sure about swapping `'a` to `'ast`. We have a convention that `'a` means AST, and `'a` lifetime is used everywhere. I assume we don't want to change that to `'ast` everywhere, right?

Reply:
The extra lifetime is unfortunate.

But from what I can see it is having a significant effect on the `RadixUIAdoptionSection.jsx` benchmark. `Transformer::new` is x2.5 faster.

The odd thing is that other changes offset that improvement. From the flamegraph, it looks like `visit_expression` doesn't get inlined any more, and that's what causes the slowdown. But I think that's a random fluke, and likely we could reverse that by just putting `#[inline]` on it.

About 10-15% of the total time on that benchmark is in setup and teardown, and I think it's worthwhile to chop that down.

So... the question is... are the lifetimes worth it for the perf improvement? Personally, I think yes. But you may have a different view.

Reply:
Actually, I *think* that if we changed the API so you pass `program` and `allocator` into `build_with_symbols_and_scopes`, instead of into `new`, then we'd be borrowing everything at the same point in time, and we'd only need one lifetime. We could get rid of `'ctx` again.

I know you were thinking of getting rid of the builder pattern, but I don't know where you go to in your thinking on that.

Reply:
> I don't know where you go to in your thinking on that

I'm on the edge of hating Rust :-)

Reply:
> I'm on the edge of hating Rust :-)

When you start hating rust, god help us all! You need some chill time, good man.

Reply:
Just to say, the perf boost on this PR was all down to adding one `#[inline]` on `visit_expression`, nothing to do with getting rid of the `Rc`s! Adding `#[inline]` was just an experiment, I would have done it in a separate PR so could see what getting rid of the `Rc`s did in isolation.